### PR TITLE
Sage 300: Completed Export Log (Parent Logic) - 1

### PR DESCRIPTION
**Pull request by anishfyle**
## Summary by CodeRabbit

- **New Features**
  - Introduced a new search and filter functionality for accounting export logs.
  - Added the ability to view detailed descriptions for expense groups.
  - Implemented date selection and pagination for export log viewing.

- **Enhancements**
  - Updated date fields to use `Date` type for better date handling.
  - Added new methods for generating URLs to view expenses in Fyle.

- **Refactor**
  - Renamed components to better reflect their functionality.
  - Removed unused components from Sage300MainModule.

- **Style**
  - Improved UI with new HTML elements and styles for better user experience.

- **Bug Fixes**
  - Fixed issues with property bindings in ExportLogFilterComponent.

- **Documentation**
  - Added comments to clarify future intentions for code enhancements.

- **Tests**
  - Created new test suites for Sage300CompleteExportLogComponent to ensure reliability.

### CODE REVIEW

**Comment by ashwin1111** [comment]
@coderabbitai review

File: src/app/core/services/common/export-log.service.ts [hunk 1]
```diff
@@ -1,9 +1,59 @@
...
+
+  private org_id: string = this.userService.getUserProfile('user').org_id;
+
+  constructor(
+    private apiService: ApiService,
+    private userService: UserService,
+    private workspaceService: WorkspaceService
+  ) { }
+
+  getReferenceType(description: Partial<ExpenseGroupDescription>): FyleReferenceType {
```
Comment (ashwin1111) [inline src/app/core/services/common/export-log.service.ts:26]: partial is not required

File: src/app/core/services/common/export-log.service.ts [hunk 2]
```diff
@@ -1,9 +1,59 @@
...
+    } else if (FyleReferenceType.EXPENSE_REPORT in description) {
+      referenceType = FyleReferenceType.EXPENSE_REPORT;
+    } else if (FyleReferenceType.PAYMENT in description) {
+      referenceType = FyleReferenceType.PAYMENT;
+    }
+
+    return referenceType;
+  }
+
+  getSkipExportLogs(limit: number, offset: number): Observable<SkipExportLogResponse> {
```
Comment (ashwin1111) [inline src/app/core/services/common/export-log.service.ts:40]: getSkippedExpenses

File: src/app/core/services/common/export-log.service.ts [hunk 3]
```diff
@@ -1,9 +1,59 @@
...
+    const workspaceId = this.workspaceService.getWorkspaceId();
+
+    return this.apiService.get(`/workspaces/${workspaceId}/fyle/expenses/`, {limit, offset});
+  }
+
+  generateFyleUrl(accountingExport: Sage300AccountingExport, referenceType: FyleReferenceType) : string {
```
Comment (ashwin1111) [inline src/app/core/services/common/export-log.service.ts:46]: type of accountingExport should be AccountingExport

File: src/app/core/services/sage300/sage300-export-log/sage300-export-log.service.ts [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+
+  private org_id: string = this.userService.getUserProfile('user').org_id;
+
+  constructor(
+    private apiService: ApiService,
+    private userService: UserService,
+    private workspaceService: WorkspaceService
+  ) { }
+
+  getAccountingExports(state: AccountingExportStatus | AccountingExportStatus.COMPLETE, limit: number, offset: number, selectedDateFilter: SelectedDateFilter | null, exportedAt: Date | void | null): Observable<AccountingExportResponse> {
```
Comment (ashwin1111) [inline src/app/core/services/sage300/sage300-export-log/sage300-export-log.service.ts:26]: move this to common service

File: src/app/core/services/sage300/sage300-export-log/sage300-export-log.service.ts [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+      const endDate = selectedDateFilter.endDate.toLocaleDateString().split('/');
+      params.start_date = `${startDate[2]}-${startDate[1]}-${startDate[0]}T00:00:00`;
+      params.end_date = `${endDate[2]}-${endDate[1]}-${endDate[0]}T23:59:59`;
+    }
+
+    if (exportedAt) {
+      params.exported_at = exportedAt;
+    }
+
+    return this.apiService.get(`/workspaces/${this.workspaceId}/fyle/expense_groups/`, params);
```
Comment (ashwin1111) [inline src/app/core/services/sage300/sage300-export-log/sage300-export-log.service.ts:44]: url is wrong

File: src/app/core/services/sage300/sage300-export-log/sage300-export-log.service.ts [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+      limit,
+      offset
+    };
+    params.state = state;
+
+    if (selectedDateFilter) {
+      const startDate = selectedDateFilter.startDate.toLocaleDateString().split('/');
+      const endDate = selectedDateFilter.endDate.toLocaleDateString().split('/');
+      params.start_date = `${startDate[2]}-${startDate[1]}-${startDate[0]}T00:00:00`;
+      params.end_date = `${endDate[2]}-${endDate[1]}-${endDate[0]}T23:59:59`;
```
Comment (ashwin1111) [inline src/app/core/services/sage300/sage300-export-log/sage300-export-log.service.ts:37]: query param might be incorrect, check API docs for usage

File: src/app/core/services/sage300/sage300-export-log/sage300-export-log.service.ts [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+  private org_id: string = this.userService.getUserProfile('user').org_id;
+
+  constructor(
+    private apiService: ApiService,
+    private userService: UserService,
+    private workspaceService: WorkspaceService
+  ) { }
+
+  getAccountingExports(state: AccountingExportStatus | AccountingExportStatus.COMPLETE, limit: number, offset: number, selectedDateFilter: SelectedDateFilter | null, exportedAt: Date | void | null): Observable<AccountingExportResponse> {
+    const params: any = {
```
Comment (ashwin1111) [inline src/app/core/services/sage300/sage300-export-log/sage300-export-log.service.ts:27]: create a type for this

File: src/app/integrations/sage300/sage300-main/sage300-export-log/sage300-complete-export-log/sage300-complete-export-log.component.html [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+
+        <div *ngIf="!filteredAccountingExports.length" class="tw-flex tw-h-3/6 tw-text-center tw-p-56-px">
+            <div class="tw-m-auto">
+                <img src="assets/illustrations/search-result-empty.svg" width="200px" hight="200px" class="tw-inline-block">
+                <div class="tw-pt-12-px">
+                    <h4 class="tw-font-500 tw-text-14-px tw-text-slightly-normal-text-color">Sorry, no results found!</h4>
+                    <p class="tw-font-400 tw-text-14-px tw-text-faded-text-color tw-pt-8-px">We could not find what you were looking for.<br>Kindly check the keywords again.</p>
+                </div>
+            </div>
+        </div>
```
Comment (ashwin1111) [inline src/app/integrations/sage300/sage300-main/sage300-export-log/sage300-complete-export-log/sage300-complete-export-log.component.html:37]: use zero state component

File: src/app/integrations/sage300/sage300-main/sage300-export-log/sage300-complete-export-log/sage300-complete-export-log.component.html [hunk 9]
```diff
@@ -0,0 +1,53 @@
...
+            <div class="tw-flex tw-h-3/6 tw-text-center tw-p-56-px tw-my-24-px">
+                <div class="tw-m-auto">
+                    <img src="assets/illustrations/integrations-zero-state.svg" width="200px" hight="200px" class="tw-inline-block">
+                    <div class="tw-pt-20-px">
+                        <h4 class="tw-font-500 !tw-text-14-px tw-text-slightly-normal-text-color">No records to show yet!</h4>
+                        <p class="tw-pt-8-px tw-font-400 !tw-text-14-px tw-text-faded-text-color">All your successful exports and their details<br>
+                            will be stored here.</p>
+                    </div>
+                </div>
+            </div>
```
Comment (ashwin1111) [inline src/app/integrations/sage300/sage300-main/sage300-export-log/sage300-complete-export-log/sage300-complete-export-log.component.html:49]: same as above

File: src/app/integrations/sage300/sage300-main/sage300-export-log/sage300-complete-export-log/sage300-complete-export-log.component.ts [hunk 10]
```diff
@@ -0,0 +1,237 @@
...
+
+  totalCount: number;
+
+  limit: number;
+
+  offset: number = 0;
+
+  currentPage: number = 1;
+
+  dateOptions: DateFilter[] = [
```
Comment (ashwin1111) [inline src/app/integrations/sage300/sage300-main/sage300-export-log/sage300-complete-export-log/sage300-complete-export-log.component.ts:34]: get options from common service, same options would be used in skip as well

File: src/app/integrations/sage300/sage300-main/sage300-export-log/sage300-complete-export-log/sage300-complete-export-log.component.ts [hunk 11]
```diff
@@ -0,0 +1,237 @@
...
+    {
+      dateRange: new Date().toLocaleDateString(),
+      startDate: new Date(),
+      endDate: new Date()
+    }
+  ];
+
+  selectedDateFilter: SelectedDateFilter | null;
+
+  presentDate = new Date().toLocaleDateString();
```
Comment (ashwin1111) [inline src/app/integrations/sage300/sage300-main/sage300-export-log/sage300-complete-export-log/sage300-complete-export-log.component.ts:59]: we're not doing anything with this, remove it off

File: src/app/integrations/sage300/sage300-main/sage300-export-log/sage300-complete-export-log/sage300-complete-export-log.component.ts [hunk 12]
```diff
@@ -0,0 +1,237 @@
...
+    private formBuilder: FormBuilder,
+    private trackingService: TrackingService,
+    private exportLogService: ExportLogService,
+    private sage300ExportLogService: Sage300ExportLogService,
+    private paginatorService: PaginatorService
+  ) { }
+
+  openExpenseinFyle(expense_id: string) {
+    const url = `${environment.fyle_app_url}/app/main/#/view_expense/${expense_id}`;
+    window.open(url, '_blank');
```
Comment (ashwin1111) [inline src/app/integrations/sage300/sage300-main/sage300-export-log/sage300-complete-export-log/sage300-complete-export-log.component.ts:83]: use window service

File: src/app/integrations/sage300/sage300-main/sage300-export-log/sage300-complete-export-log/sage300-complete-export-log.component.ts [hunk 13]
```diff
@@ -0,0 +1,237 @@
...
+  constructor(
+    private formBuilder: FormBuilder,
+    private trackingService: TrackingService,
+    private exportLogService: ExportLogService,
+    private sage300ExportLogService: Sage300ExportLogService,
+    private paginatorService: PaginatorService
+  ) { }
+
+  openExpenseinFyle(expense_id: string) {
+    const url = `${environment.fyle_app_url}/app/main/#/view_expense/${expense_id}`;
```
Comment (ashwin1111) [inline src/app/integrations/sage300/sage300-main/sage300-export-log/sage300-complete-export-log/sage300-complete-export-log.component.ts:82]: move logic to common service

File: src/app/integrations/sage300/sage300-main/sage300-export-log/sage300-complete-export-log/sage300-complete-export-log.component.ts [hunk 14]
```diff
@@ -0,0 +1,237 @@
...
+    window.open(url, '_blank');
+  }
+
+  public filterTableChange(event: any) {
```
Comment (ashwin1111) [inline src/app/integrations/sage300/sage300-main/sage300-export-log/sage300-complete-export-log/sage300-complete-export-log.component.ts:86]: handleSimpleSearch()

File: src/app/integrations/sage300/sage300-main/sage300-export-log/sage300-complete-export-log/sage300-complete-export-log.component.ts [hunk 15]
```diff
@@ -0,0 +1,237 @@
...
+      const expenseType = group.expenseType ? group.expenseType : '';
+      const referenceNumber = group.referenceNumber ? group.referenceNumber : '';
+
+      return (
+        employeeName.toLowerCase().includes(query) ||
+        employeeID.toLowerCase().includes(query) ||
+        expenseType.toLowerCase().includes(query) ||
+        referenceNumber.toLowerCase().includes(query)
+      );
+    });
```
Comment (ashwin1111) [inline src/app/integrations/sage300/sage300-main/sage300-export-log/sage300-complete-export-log/sage300-complete-export-log.component.ts:101]: move logic to common service

File: src/app/integrations/sage300/sage300-main/sage300-export-log/sage300-complete-export-log/sage300-complete-export-log.component.ts [hunk 16]
```diff
@@ -0,0 +1,237 @@
...
+  }
+
+  dateFilter(event: any): void {
+    this.isLoading = true;
+    this.isDateSelected = true;
+    this.selectedDateFilter = event.value;
+    this.getAccountingExports(this.limit, this.offset);
+  }
+
+  dropDownWatcher() {
```
Comment (ashwin1111) [inline src/app/integrations/sage300/sage300-main/sage300-export-log/sage300-complete-export-log/sage300-complete-export-log.component.ts:127]: where is this called?

File: src/app/integrations/sage300/sage300-main/sage300-export-log/sage300-complete-export-log/sage300-complete-export-log.component.ts [hunk 17]
```diff
@@ -0,0 +1,237 @@
...
+    if (this.exportLogForm.controls.dateRange.value !== this.dateOptions[3].dateRange) {
+      this.isCalendarVisible = false;
+    } else {
+      this.isCalendarVisible = true;
+    }
+  }
+
+  showCalendar(event: Event) {
```
Comment (ashwin1111) [inline src/app/integrations/sage300/sage300-main/sage300-export-log/sage300-complete-export-log/sage300-complete-export-log.component.ts:135]: where is this called?

File: src/app/integrations/sage300/sage300-main/sage300-export-log/sage300-complete-export-log/sage300-complete-export-log.component.ts [hunk 18]
```diff
@@ -0,0 +1,237 @@
...
+    event.stopPropagation();
+    this.isCalendarVisible = true;
+  }
+
+  getDates() {
```
Comment (ashwin1111) [inline src/app/integrations/sage300/sage300-main/sage300-export-log/sage300-complete-export-log/sage300-complete-export-log.component.ts:140]: getDates

File: src/app/integrations/sage300/sage300-main/sage300-export-log/sage300-complete-export-log/sage300-complete-export-log.component.ts [hunk 19]
```diff
@@ -0,0 +1,237 @@
...
+    this.dateOptions[3].dateRange = this.exportLogForm.value.start[0].toLocaleDateString() + '-' + this.exportLogForm.value.start[1].toLocaleDateString();
+    this.dateOptions[3].startDate = this.exportLogForm.value.start[0];
+    this.dateOptions[3].endDate = this.exportLogForm.value.start[1];
+    this.presentDate = this.dateOptions[3].dateRange;
+    this.exportLogForm.controls.dateRange.patchValue(this.dateOptions[3]);
+  }
+
+  getAccountingExports(limit: number, offset:number) {
```
Comment (ashwin1111) [inline src/app/integrations/sage300/sage300-main/sage300-export-log/sage300-complete-export-log/sage300-complete-export-log.component.ts:148]: private func, return type missing

File: src/app/integrations/sage300/sage300-main/sage300-export-log/sage300-complete-export-log/sage300-complete-export-log.component.ts [hunk 20]
```diff
@@ -0,0 +1,237 @@
...
+          employee: [accountingExport.expenses[0].employee_name, accountingExport.description.employee_email],
+          expenseType: accountingExport.fund_source === 'CCC' ? 'Corporate Card' : 'Reimbursable',
+          fyleReferenceType: '',
+          referenceNumber: referenceNumber,
+          exportedAs: accountingExport.type,
+          fyleUrl: this.exportLogService.generateFyleUrl(accountingExport, referenceType),
+          integrationUrl: accountingExport.export_url,
+          expenses: accountingExport.expenses
+        });
+      });
```
Comment (ashwin1111) [inline src/app/integrations/sage300/sage300-main/sage300-export-log/sage300-complete-export-log/sage300-complete-export-log.component.ts:181]: logic should be moved to common model

File: src/app/integrations/sage300/sage300-main/sage300-export-log/sage300-complete-export-log/sage300-complete-export-log.component.ts [hunk 21]
```diff
@@ -0,0 +1,237 @@
...
+          referenceNumber = accountingExport.expenses[0].expense_number;
+        } else if (referenceType === FyleReferenceType.PAYMENT) {
+          referenceNumber = accountingExport.expenses[0].payment_number;
+        }
+
+        accountingExports.push({
+          exportedAt: accountingExport.exported_at,
+          employee: [accountingExport.expenses[0].employee_name, accountingExport.description.employee_email],
+          expenseType: accountingExport.fund_source === 'CCC' ? 'Corporate Card' : 'Reimbursable',
+          fyleReferenceType: '',
```
Comment (ashwin1111) [inline src/app/integrations/sage300/sage300-main/sage300-export-log/sage300-complete-export-log/sage300-complete-export-log.component.ts:174]: missing

File: src/app/integrations/sage300/sage300-main/sage300-export-log/sage300-complete-export-log/sage300-complete-export-log.component.ts [hunk 22]
```diff
@@ -0,0 +1,237 @@
...
+        if (referenceType === FyleReferenceType.EXPENSE) {
+          referenceNumber = accountingExport.expenses[0].expense_number;
+        } else if (referenceType === FyleReferenceType.PAYMENT) {
+          referenceNumber = accountingExport.expenses[0].payment_number;
+        }
+
+        accountingExports.push({
+          exportedAt: accountingExport.exported_at,
+          employee: [accountingExport.expenses[0].employee_name, accountingExport.description.employee_email],
+          expenseType: accountingExport.fund_source === 'CCC' ? 'Corporate Card' : 'Reimbursable',
```
Comment (ashwin1111) [inline src/app/integrations/sage300/sage300-main/sage300-export-log/sage300-complete-export-log/sage300-complete-export-log.component.ts:173]: enum

File: src/app/integrations/sage300/sage300-main/sage300-export-log/sage300-complete-export-log/sage300-complete-export-log.component.ts [hunk 23]
```diff
@@ -0,0 +1,237 @@
...
+      }
+      accountingExportResponse.results.forEach((accountingExport: Sage300AccountingExport) => {
+        const referenceType: FyleReferenceType = this.exportLogService.getReferenceType(accountingExport.description);
+        let referenceNumber: string = accountingExport.description[referenceType];
+
+        if (referenceType === FyleReferenceType.EXPENSE) {
+          referenceNumber = accountingExport.expenses[0].expense_number;
+        } else if (referenceType === FyleReferenceType.PAYMENT) {
+          referenceNumber = accountingExport.expenses[0].payment_number;
+        }
```
Comment (ashwin1111) [inline src/app/integrations/sage300/sage300-main/sage300-export-log/sage300-complete-export-log/sage300-complete-export-log.component.ts:168]: move to separate func and model

**Review by ashwin1111 — APPROVED** [review]

**Comment by anishfyle** [comment]
@coderabbitai review

**Pull request merged**